util/opentelemetry-util-http/src/opentelemetry/util/http/httplib.py

```diff
@@ -0,0 +1,175 @@
+# Copyright The OpenTelemetry Authors
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""
+This library provides functionality to enrich HTTP client spans with IPs. It does
+not create spans on its own.
+"""
+
+import contextlib
+import http.client
+import logging
+import socket  # pylint:disable=unused-import # Used for typing
+import typing
+from typing import Collection
+
+import wrapt
+
+from opentelemetry import context
+from opentelemetry.instrumentation.instrumentor import BaseInstrumentor
+from opentelemetry.instrumentation.utils import unwrap
+from opentelemetry.semconv.trace import SpanAttributes
+from opentelemetry.trace.span import Span
+
+_STATE_KEY = "httpbase_instrumentation_state"
+
+logger = logging.getLogger(__name__)
+
+
+class HttpClientInstrumentor(BaseInstrumentor):
```

Comment:
I'm not too sure how this is supposed to be used by the user. Is this `Instrumentor` instantiated as well along side `RequestsInstrumentor`? Or is `RequestsInstrumentor` supposed to extend this class?

Reply:
Oh, actually now that you say it, this was the reason why it was in a separate package...

This is not supposed to be used by the user at all. Instead it should just be loaded by the instrumentation infrastructure and then it does its thing behind the scentes. The requests and urllib3 instrumentations are the actual users of this code, since they register their wish to have the IP set onto a span which this instrumentation fulfills once the standard library httplib opens a socket.

Reply:
I think I have to add an entrypoint to the setup.cfg of util-http as well as the other instrumentation boilerplate. Do you think it would be better to keep it as separate package?

Reply:
I don't mind having an entrypoint for client libraries, unless you feel strongly having another package. I personally don't like having many packages :)

Reply:
> How come we need to create a new instrumentor for this feature?

Because this instruments the standard library httplib.client. I could put it in the urllib3 instrumentation, but if we ever want to instrument another library based on httplib.client (e.g. there is a urllib instrumentation here that could also be enhanched with this), then where should we put it if not in a separate, independent instrumentor?

> Do all client libraries have a "send" method?

No, just those based on httplib.client, which appies to requests and urllib3 at least.

Reply:
So maybe the point to emphasize here: This is *not* a completely generic HTTP IP detection (at least yet). It is a instrumentation for httplib.client except that it does not create spans on its own but cooperates with other instrumentations that instrument libs which use httplib.client internally.

Reply:
> If not, maybe we can just have trysetip and set_ip_on_next_http_connection as utility functions, and decorate the request functions directly?

Sorry, I don't follow this. What do you mean by request functions I should decorate? The IP cannot be obtained from the API functions we usually instrument since there is no access to the underlying socket, or accessing it would be very complicated.

Reply:
> The requests and urllib3 instrumentations are the actual users of this code,

Sorry I might be missing something. I'm not understanding how the two instrumentations actually use `HttpClientInstrumentor`? Even the tests have:

```
HttpClientInstrumentor().instrument()
URLLib3Instrumentor().instrument()
```
so it's a bit unclear in how these are supposed to be used.

Reply:
> Currently, it looks like only set_ip_on_next_http_connection is being used in the instrumentations.

Exactly, this is the *only* feature this new instrumentation provides. It sets net.peer.ip to the actually connected-to IP (vs some separate DNS lookup which will usually not yield the same IP again)

Reply:
> Even the tests have:
> ```
> HttpClientInstrumentor().instrument()
> URLLib3Instrumentor().instrument()
> ```

Yes. The tests need to call `HttpClientInstrumentor().instrument()` manually since no distribution (is it called that way) is loaded that auto-enables the instrumentation. `URLLib3Instrumentor().instrument()` cannot call `HttpClientInstrumentor().instrument()` IIUC, as that would lead to double instrumentation. Also, I wanted to have the feature optional. If users disable the HttpClientInstrumentor (not sure if this is possible today), they won't get net.peer.ip and potential associated overhead and bugs.

Reply:
Just confirming here. As a user, if I manually instrument code, and load the `RequestsInstrumentor`, I will also need to manually load the `HttpClientInstrumentor` myself correct?

Would it be possible to call `trysetip` from the requests/urlib3 code and skip the additional instrumentor? Or is the problem that those libraries don't have access to the underlying socket to pull IP information from it?

Reply:
Yes, the main reason was that its not easy to get the underlying socket. I did not check for every library separately, in some it may be possible to use some undocumented properties or instrument some constructors to get to the socket. Instrumenting httplib.client, although also awkward, seemed like the most clean way.

Reply:
Got it, that makes sense. Thanks!

Reply:
@codeboten To follow up here: Is manually using the instrumentors common enough that it would make sense to add a hint to use HttpClientInstrumentor to the READMEs of the requests and urllib3 instrumentations?